Implement auto updater for Windows and Mac

yuya-oc opened the pull request:
Before submitting, please confirm you've
 - [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
Implement auto updater.

This is a first prototyping of auto-updater. At this time, this PR only considers the Windows installer.

NSIS installer is introduced. It integrates 32-bit and 64-bit installers into one installer file. And the new installation folder becomes `%USERPROFILE%\AppData\Local\Programs\mattermost` (Per user install). We can also choose `C:\Program Files` (Per machine install) by modifying electron-builder settings.

<img width="358" alt="updater-dialog" src="https://user-images.githubusercontent.com/1412057/30032314-31557946-91d0-11e7-9a17-be17eb8667f9.png">


**Issue link**
#15 

**Test Cases**
I prepared a version, v9.9.9 for testing: https://circleci.com/gh/yuya-oc/desktop/366#artifacts

A. simple case with CircleCI

1. Install the app.
2. Edit `%USERPROFILE%\AppData\Local\Programs\mattermost\resources\app-update.yml`
    - Set `'https://366-67276967-gh.circle-artifacts.com/0/tmp/circle-artifacts.0utXb6g/'` to the value of `url`
3. The app will check updates to CircleCI URL.

B. manual testing for various cases.
1. Prepare a http server to locate `latest.yml` and `mattermost-setup-9.9.9.exe`.
    Example:  [`http-server`](https://www.npmjs.com/package/http-server)
    1. Install Node.js.
    2. Execute `npm install -g http-server`
    3. Execute `http-server -p 8081` under the folder where artifacts are located.
2. The app will check updates to `http://localhost:8081/latest.yml`.
3. You can change the test version and release date by editing `latest.yml`.

**Additional Notes**
https://circleci.com/gh/yuya-oc/desktop/365#artifacts

jasonblais commented:
Wow wow wow!!!

Thanks! Will begin the review soon (we're cutting the release candidate for the next server release tomorrow so it might be a couple of days)

MusikPolice commented on src/browser/components/UpdaterPage.jsx:
```diff
@@ -0,0 +1,53 @@
+  return (
+    <div>
+      <Navbar>
+        <h1 className='UpdaterPage-heading'>{'New update is available'}</h1>
```
Do these strings need to be translated into the users' language?

yuya-oc replied:
Not necessary because the desktop app doesn't have any localization system for now.

MusikPolice commented on src/main/autoLaunch.js:
```diff
@@ -0,0 +1,17 @@
+const AutoLaunch = require('auto-launch');
+
+async function upgradeAutoLaunch() {
+  if (process.platform === 'darwin') {
```
Darwin is MacOS, right? What does this do on Linux platforms? If it's only supported on Windows, I wonder if a whitelist approach might be best here.

yuya-oc replied:
Yeah, `darwin` means macOS. On Linux, this function should behave as well as Windows: Reconfigure auto-start for the current version.

MusikPolice commented on src/main/autoUpdater.js:
```diff
@@ -0,0 +1,101 @@
+}
+
+function isUpdateApplicable(appState, updateInfo) {
+  if (appState.updateCheckedDate.value - (new Date(updateInfo.releaseDate)).value < interval48hours) {
```
Are these two dates always in a fixed time zone? Like are they both UTC dates? If either is based on the user's timezone, then we may not actually wait 48 hours.

yuya-oc replied:
UTC is always used.

- updateCheckedDate: [Date.prototype.toISOString()](https://github.com/mattermost/desktop/pull/582/files/972a233ed49bceea2d0ed419c4aa0cafd60adb7f#diff-f86667a718f2bfb29e313e6c1f73a02eR13)
- releaseDate: I can't show in source code. But I confirmed that the date uses UTC ('Z' in ISO format)

jasonblais commented:
> We can also choose C:\Program Files (Per machine install) by modifying electron-builder settings.

Does this mean we as in "us" or can the admin modify this in anyway (I assume it's "us" but thought to ask).

As for testing, wondering if there's anything I can test (except review the UI you posted above) until we have a local http server of some kind

yuya-oc commented:
Yeah, it means "us". We can choose one of them:

- Per user (under `AppData`) (current)
- Per machine (under `C:\Program Files`)
- Choose one of them in install wizard

Whichever we choose one of them, we can add the option to select install location by adding install wizard.

This should be discussed at #491, but I wanted to tell a little here.

yuya-oc commented:
Updated the artifacts link and test cases. You would be able to see the dialog after 48 hours by following the steps A.

jasonblais commented:
@yuya-oc Let's not merge #582  until PRs contributed by @csduarte are merged to avoid conflicts for those PRs 

https://github.com/mattermost/desktop/pulls/csduarte

yuya-oc commented:
@jasonblais Okay, I will continue to work on this branch.

jasonblais commented:
Thanks @yuya-oc! I'll also get feedback from others about our approach

yuya-oc commented:
Rebased and added a feature for Windows zip. When using zip packages, "Install Update" button becomes "Download Update" button and it opens https://about.mattermost.com/download/#mattermostApps in user's default browser.

yuya-oc commented:
Rebased: https://circleci.com/gh/yuya-oc/desktop/500#artifacts

@jasonblais To test Mac, we need code-signing for newer app (v9.9.9: https://circleci.com/gh/yuya-oc/desktop/501#artifacts) Would you help this? (automation is not needed for now)

jasonblais commented:
@yuya-oc Awesome! Thanks. I've set time for myself to test this PR thoroughly, and planning to ask other team members to help as well.

I'll ask Joram or Jonathan to help code-signing for the Mac

jasonblais commented:
@yuya-oc I installed the Windows .exe file and edited the URL for `app-update.yml` as `https://366-67276967-gh.circle-artifacts.com/0/tmp/circle-artifacts.0utXb6g/`.

I'll wait for 48 hours to see if the update prompt comes up Tuesday evening.

Few initial questions/thoughts:

1) Do we have a flag somewhere to turn off auto-updates by the admin (e.g. a customer deploying a custom fork of the app, or a package manager like Chocoleatey distributing their own package)?
  - if not, would this be added to BuildConfig.json, perhaps?
2) Can you help update NOTICE.txt with new dependancies?
3) Can you help edit the upgrade prompt interval from 48 hours to 5 minutes for testing purposes? It'll make it faster to test all the different packages, as I'm planning to set aside time for Monday/Tuesday to go through all 4 variations for Windows, and one for Mac, and hopefully get other people to test it out too.
4) Can you help add a "Check for Updates..." option in the "Help" menu bar, as described in [slides 13 and 14 of the spec](https://docs.google.com/presentation/d/1BmH5gGtdhijHRfv46_uhgl_0kwT5qOI8_5y2HNfNpqA/edit#slide=id.g1f0e8df706_0_23)
5) How much work will be involved for adding the auto-update prompt for Linux, similar to what's implemented for Windows .zip files? (Just so I know)

jasonblais commented:
Also, just want to reiterate a big thanks for your help on this PR. It'll be a great feature to add for v4.0.

My aim is to get all testing and feedback summarized here in the next few days, and prepare to merge this PR within the next two weeks for v4.0 🎉

yuya-oc commented:
@jasonblais Sorry, probably I had mistake when rebasing. Please stop test for a moment.

yuya-oc commented:
Updated: https://circleci.com/gh/yuya-oc/desktop/502#artifacts

1. Now I'm considering several options.
  - Package maintainers: By adding an extra file (app-updater-config.json) into packages, they can override auto-updater parameters.
  - Admins: They can override parameters via pre-configuration.
2. Okay, I'll do later.
3. https://circleci.com/gh/yuya-oc/desktop/503#artifacts
4. It was missing due to my mistake. It should appear at current artifacts.
5. Prototype is ready. However it's written from zero. So for now, I would like to focus on Windows and Mac.

jasonblais commented:
1 - Would having the options in buildConfig.json be an option? Since it's used to configure the app before building it?
5 - Sounds good 👍

jasonblais commented:
And thanks for the changes! I'll spend time today and tomorrow to test this out. Big thanks!

yuya-oc commented:
> 1 - Would having the options in buildConfig.json be an option? Since it's used to configure the app before building it?

Auto updater parameters such as URL are defined in `electron-builder.json`. Package maintainers can edit it by themselves. What about admins? They would need to build the app from source.

jasonblais commented:
Thans @yuya-oc. I think the main use case for admins would be when they're running their own versions of the Mattermost app (e.g. have forked us), and hence want to disable it, or somehow send the notification out themselves for their own releases.

Or is there a use case for admins I'm not thinking about? Would one case be when the app is distributed silently to their end users?

Definitely important to think this through so we don't fire update notifications that conflict with the admin set up.

csduarte commented:
@yuya-oc the ability to turn this on/off at choice will be important. Eventually, we will probably want to use it, but we would have to test. Our fleet is 40,000+ machines with all forms of OS's. Something like this will need to best tested and also security is a concern.

yuya-oc commented:
@jasonblais The use case I thought is that they use our prebuilt packages. However it would be confusing point like #483.

@csduarte Yeah, that's understandable. At least it should be one of `buildConfig` parameters.

jasonblais commented:
@yuya-oc Having it in `buildConfig.json` for package managers (and anyone else packaging the apps themselves. and distributing to end users), and in the pre-config file for admins seems okay to me.

1 - For package maintainers: We can add it to either `buildConfig.json` or have an `app-updater-config.json` file. 

`buildConfig.json` feels simpler and more obvious to me, since that's where we have the other build settings, but I don't have a strong opinion on this. We could check with Hans or William who are maintaining the Chocolatey packages and see if they have an opinion (`hans0` and `wget` on pre-release.mattermost.com, respectively).

2 - For admins, you bring a good point about using prebuild packages, and distributing them to users. For that, including it in the pre-config file makes sense.

yuya-oc commented:
Fixed an unexpected exception and added `enableAutoUpdater` to buildConfig. https://circleci.com/gh/yuya-oc/desktop/515

5 min: https://circleci.com/gh/yuya-oc/desktop/517#artifacts
v9.9.9: https://circleci.com/gh/yuya-oc/desktop/516#artifacts

@jasonblais

1 - Surely buildConfig is obvious. However to edit it makes build steps difficult because we need to build the app in various config at once. So I added `app-updater-config.json` to control the updater.

2 - Yeah, I will add after pre-config is ready.

jasonblais commented:
@yuya-oc Great, I will test this.

1 & 2 - sounds good to me. We should document all this carefully via https://github.com/mattermost/desktop/issues/648

jasonblais commented:
Have also asked Joram to help with code-signing of the Mac apps. If it fails again, we'll try to get you access to the credentials to troubleshoot the issue.

jasonblais commented:
Just an update: Joram completed code-signing last week and have been testing. Will share feedback once finished.

jasonblais commented:
Left a note for you on pre-release,

yuya-oc commented:
I fixed some problem.

a. "Remind me later" not working.

b. The app installs the update when quitting the app even when using "Remind me later".
  So I stopped auto-download of updater, and added a progress bar for downloading.

<img width="358" alt="progress" src="https://user-images.githubusercontent.com/1412057/34000062-954faff8-e12f-11e7-9abe-cff493b34278.png">

5 min: https://circleci.com/gh/yuya-oc/desktop/530#artifacts
v9.9.9: https://circleci.com/gh/yuya-oc/desktop/531#artifacts

jasonblais commented:
@yuya-oc 

@amyblais and I tried the latest artifact (both 530 and 531) and couldn't get it to work. Amy tried on Windows, I tried on Mac.

E.g. after installing the app and clicking "Check for Updates" does nothing.

Wondering if there's something we're doing incorrectly?

yuya-oc commented:
@jasonblais Did you follow the instruction A-2?
 `%USERPROFILE%\AppData\Local\Programs\mattermost\resources\app-update.yml` should be below.

```yml
provider: generic
url: 'https://531-67276967-gh.circle-artifacts.com/0/tmp/circle-artifacts.OL3096N/'
```

jasonblais commented:
@yuya-oc I'll try to re-test this next week

jasonblais commented:
@yuya-oc If we added the apps to Mac Apple Store and Windows Store, and created an official release repo for Linux, then users would get automatically notified by the updates, correct?

yuya-oc commented:
@jasonblais In current implementation, no.

- Mac App Store, Windows Store
Usually notifying/upgrading are managed by the store itself. So probably we don't have to take care about these cases.

- Official release repo for Linux
Possibly we should notify updates. System's package manager would notify though, user's might not check it.

jasonblais commented:
@yuya-oc Makes sense. I'm just wondering that if we did add the apps to Mac App Store and Windows Store, at least a portion of users would download it and get subsequent updates from the store.

Maybe that's something we should consider prioritizing while we work through this PR. Thoughts?

yuya-oc commented:
@jasonblais For now, no. To submit to stores, we need to build specialized packages. We should consider that first.

jasonblais commented:
@yuya-oc Now that the build process and code signing are automated, can you help generate new test builds based on the new job? 

That way we can test this feature with the new build process.

esethna commented:
@yuya-oc just following up, when you have a chance can we generate new test builds for this?

mcallaghan-sandvine commented:
If you need Linux testing let me know.

esethna commented:
@yuya-oc a couple notes from testing:

1. What is expected for the action button from this dialog? On Windows it says "Download Update" and takes me to the mattermost apps download page: https://mattermost.com/download/#mattermostapps. But on Mac it says "Install Update" and then I get a progress bar for the download I'm assuming.

![image](https://user-images.githubusercontent.com/14333569/46756285-52b2d080-cc84-11e8-82bd-99edaa9cbe0a.png)

2. For the download progress it looks like the white XX% text needs to be moved to the right slightly to add some buffer space:

![image](https://user-images.githubusercontent.com/14333569/46756438-b2a97700-cc84-11e8-8a78-b6ffaf941753.png)

3. Can we add a "Cancel" button on the update dialog once the download starts? I had to quit and restart the app since it was stuck at 0% and I couldn't get out of the dialog.

yuya-oc commented:
@esethna Thanks for feedback.

1. Which version of Widows app did you use, zip or installer? When using zip version, auto-updater can't work. So it says "Download Update" and opens the link for the download page.

2. [By adding `min-width` parameter](https://getbootstrap.com/docs/3.3/components/#progress-label), we can get readable letters. Is it acceptable?  <img width="384" alt="zero_percent_progress_bar" src="https://user-images.githubusercontent.com/1412057/46961361-91f57d00-d0db-11e8-9fcb-90ddf195ff73.png">

3. Probably we should do that. I'll take a look.

esethna commented:
1. Ah okay that's why, I was using the ZIP. So when the desktop app is installed via the Mattermost download page by users it is expected that the auto-updater will show the progress bar and install the update from this modal, as I experienced on my mac? 
2. Yes looks good thanks.
3. Cool, thanks!

yuya-oc commented:
Hi @esethna. I updated the PR. Now it's ready to test 0% of progress bar and the cancel button.

Please get artifacts from jenkins: https://build.mattermost.com/job/md/job/desktop-pulls/view/change-requests/job/PR-582/53/

> Ah okay that's why, I was using the ZIP. So when the desktop app is installed via the Mattermost download page by users it is expected that the auto-updater will show the progress bar and install the update from this modal, as I experienced on my mac?

That's right.

esethna commented:
@yuya-oc retested the build artifacts you linked. Looks like I'm getting this modal stating my app is up to date after waiting 5 minutes or when checking for updates:

![image](https://user-images.githubusercontent.com/14333569/47522846-d8717700-d853-11e8-9a30-552ebd2d6971.png)
_______

Also, given we are close to being ready to merge this, do we have to make any changes to get this ready for production? And is there a way for our QA team test this on an RC after merge?

yuya-oc commented:
@esethna Well, it's weird. I got the modal to notify new release. Would you make sure the app version when you start the app?

esethna commented:
As per new review policy, adding @hmhealey to help do a pass

hmhealey commented on resources/windows/app-updater-config.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "notifyOnly": true
```
Is this first version just going to notify the user that there's updates, or are we going to supply a different config later on?

Also, where is it going to check for these updates?

yuya-oc replied:
This static config file is only bundled into zip version of Windows app because zip version can't use auto-updater. It should be read-only and change a button in the dialog as esethna had reviewed. Then the button will open https://mattermost.com/download/ .

![notifyonly-dialog](https://user-images.githubusercontent.com/14333569/46756285-52b2d080-cc84-11e8-82bd-99edaa9cbe0a.png)

The updater will check `latest.yml` and `latest-mac.json` under a base url. The base url is defined in `electron-builder.json`.

https://github.com/mattermost/desktop/blob/849800a80fb4c44242c1ddab824ccf9ba451890c/electron-builder.json#L4

The base url should be `https://releases.mattermost.com/desktop/` finally.

hmhealey replied:
Ah, I knew it was somewhere. So it'll look at `http://desktopupdate.mattermost.com/latest.yml` and `http://desktopupdate.mattermost.com/latest-mac.json` to get those updates. Could we switch those to use https?

Also, this doesn't really matter, but is there a reason you chose to mix yml and json files for those two?

yuya-oc replied:
Sure, We can use https.

Mixing yml and json is derived from `electron-builder` spec. It generates them at build-time, so we can't control.

hmhealey replied:
Makes sense that that's why we switch between the two. It's a bit odd, but it's not a problem.

Using https is important though since it would ensure that the file is actually coming from us and the request isn't being intercepted by someone else

esethna commented:
Thanks @yuya-oc! @hmhealey please help review the changes?

crspeller commented on src/browser/components/UpdaterPage/UpdaterPage.stories.jsx:
```diff
@@ -0,0 +1,53 @@
+onClickRemind: propTypes.func.isRequired,
+onClickSkip: propTypes.func.isRequired,
+*/
+const appName = 'Storybook App';
```
What's this about?

yuya-oc replied:
Just for [storybook](https://storybook.js.org/) previewing. It appears as application name as a part of dialog sentences. To white labeling the app (e.g. uchat), the app name is injected to the component as a prop.

You can see this name at http://localhost:9001 while running `npm run storybook`.

crspeller commented on src/browser/components/UpdaterPage/UpdaterPage.stories.jsx:
```diff
@@ -0,0 +1,53 @@
+      notifyOnly={false}
+      isDownloading={false}
+      progress={0}
+      onClickInstall={action('clicked install')}
```
Why doesn't this match anything? Is it supposed to be `click-install`?

yuya-oc replied:
Yes. On the actual app, `click-install` ipc message is sent from the updater dialog to the main process in order to invoke the function of electron-updater.

esethna commented:
@crspeller feedback above ^

crspeller commented:
@yuya-oc Looking for answer to my question so we can get stuff resolved and merged.

yuya-oc commented:
Thanks @wget. Please let me confirm.

## Windows
Should v4.2 app always use "Download Update" button for the dialog? The button opens a web page, https://mattermost.com/download/

![download](https://user-images.githubusercontent.com/14333569/46756285-52b2d080-cc84-11e8-82bd-99edaa9cbe0a.png)

## macOS
Agree. In addition, recent `brew cask` package manager has a parameter to know whether the app has auto-updater. So there would be no conflicts.

## Linux
Agree. However, currently there is no implementation for Linux because I had thought we need to finalize Windows/macOS solution first. Can we continue to discuss on another issue?

crspeller commented:
@yuya-oc I don't see our public key anywhere in this PR, how does the signature verification work?

wget commented:
@yuya-oc 
> ## Windows
> Should v4.2 app always use "Download Update" button for the dialog? The button opens a web page, https://mattermost.com/download/

If this is just a link to a web page. No issue.

> ## macOS
> Agree. In addition, recent `brew cask` package manager has a parameter to know whether the app has auto-updater. So they would not conflict.

Ok. Understood.

> 
> ## Linux
> Agree. However, currently there is no implementation for Linux because I had thought we need to finalize Windows/macOS solution first. Can we continue to discuss on another issue?

Ok. Understood. GNU/Linux is indeed low priority now because its the only OS having excellent package managers. End users won't suffer from a missing auto updater (especially on Arch Linux, since the packages are well maintained <3).

wget commented:
> @yuya-oc I don't see our public key anywhere in this PR, how does the signature verification work?

Indeed this needs to get sorted out. This is a feature I need for msi installer as well. Please provide info how you are singing the applications. So that I can follow the same path for my msi.

This is what happens with non signed msi installers that have been downloaded from the web:

![screenshot_20181105_092311](https://user-images.githubusercontent.com/2997662/48640585-649c3780-e9d7-11e8-96cc-538651dedffc.png)

yuya-oc commented:
Yeah, “Download Update” button behaves like a link. The download page is opened in their default web browser.

@wget @crspeller In my understanding, signature verification is done by Windows itself. So what we need to do in ourselves is verify checksum. Codesining is done by our jenkins, https://build.mattermost.com. So the script to sign exe is included in the private repository (shared pipeline), not in this repository. The script gets private files from Mattermost’s Amazon S3 which stores certificates.

The script manually signs the code with osslsigncode. However electron-builder has a mechanism to do code-signing while building the app. We move to AppVeyor, so I believe we can use it.
https://www.electron.build/code-signing#windows

wget commented:
@yuya-oc Thanks. Super. I will add these pieces of info together and will try to figure out to add the code signing in the AppVeyor pipeline.

yuya-oc commented:
Updated with "Download Update" (notifyOnly) configuration for Windows.

esethna commented:
@yuya-oc can you please address the build issue when you can ^ Then I think we are good to merge this? 

This should be the last PR before we cut the RC for 4.2?

yuya-oc commented:
@esethna Now the error is resolved, so I think this is ready to merge.

I'll make another PR. That's the final PR for v4.2: https://github.com/mattermost/desktop/pull/841#issuecomment-428180817

yuya-oc commented:
Merged the master branch. For now I merge this branch to suppress more conflicts. We can disable the feature when we need.
